### ISSUE TITLE
[BUG] Fail to build docker image for cpu environment

### ISSUE THREAD

**Issue by hsm207** [microsoft/recommenders#1026]
### Description
Building the docker image for cpu environment fails when building xlearn.

### In which platform does it happen?
 Ubuntu 18.04.3 LTS

### How do we replicate the issue?
Run the following commands (taken from [here](https://github.com/microsoft/recommenders/blob/master/docker/README.md))

`DOCKER_BUILDKIT=1 docker build -t recommenders:cpu --build-arg ENV="cpu" .`

### Expected behavior (i.e. solution)
Build completes successfully

### Other Comments

This is the output from the build command:

```
#11 369.3 Failed to build xlearn
#11 369.3 Installing collected packages: distro, dotnetcore2, azureml-dataprep-native, fusepy, azureml-dataprep, PyJWT, adal, oauthlib, requests-oauthlib, isodate, msrest, msrestazure, azureml-train-restclients-hyperdrive, azure-common, azure-mgmt-authorization, pyasn1, ndg-httpsclient, websocket-client, docker, azure-mgmt-resource, azure-graphrbac, jeepney, SecretStorage, azure-mgmt-keyvault, jsonpickle, ruamel.yaml, contextlib2, azure-mgmt-storage, azure-mgmt-containerregistry, jmespath, pathspec, azureml-core, applicationinsights, azureml-telemetry, azureml-train-core, azureml-train, azureml-pipeline-core, azureml-pipeline-steps, azureml-pipeline, azureml-contrib-notebook, azureml-widgets, azureml-tensorboard, azureml-sdk, azure-nspkg, azure-storage, colorama, wheel, argcomplete, humanfriendly, bcrypt, pynacl, paramiko, antlr4-python3-runtime, portalocker, azure-cli-nspkg, azure-cli-telemetry, tabulate, knack, azure-cli-core, azure-mgmt-cosmosdb, typed-ast, toml, regex, appdirs, black, category-encoders, dataclasses, pymongo, networkx, hyperopt, idna, greenlet, gevent, itsdangerous, flask, locustio, memory-profiler, nbconvert, pydocumentdb, pymanopt, xlearn, schema, coverage, simplejson, PythonWebHDFS, json-tricks, nni
#11 369.3   Found existing installation: wheel 0.33.6
#11 369.3     Uninstalling wheel-0.33.6:
#11 369.3       Successfully uninstalled wheel-0.33.6
#11 369.3   Found existing installation: idna 2.8
#11 369.3     Uninstalling idna-2.8:
#11 369.3       Successfully uninstalled idna-2.8
#11 369.3   Found existing installation: nbconvert 5.6.1
#11 369.3     Uninstalling nbconvert-5.6.1:
#11 369.3       Successfully uninstalled nbconvert-5.6.1
#11 369.3     Running setup.py install for xlearn: started
#11 369.3     Running setup.py install for xlearn: finished with status 'error'
#11 369.3
#11 369.3 CondaEnvException: Pip failed
#11 369.3
#11 369.3
#11 369.3 Pip subprocess error:
#11 369.3   ERROR: Command errored out with exit status 1:
#11 369.3    command: /root/conda/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-_5jp1bcj/xlearn/setup.py'"'"'; __file__='"'"'/tmp/pip-install-_5jp1bcj/xlearn/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-aconjv2v --python-tag cp36
#11 369.3        cwd: /tmp/pip-install-_5jp1bcj/xlearn/
#11 369.3   Complete output (47 lines):
#11 369.3   /root/conda/lib/python3.6/site-packages/setuptools/dist.py:476: UserWarning: Normalizing '0.40.a1' to '0.40a1'
#11 369.3     normalized_version,
#11 369.3   running bdist_wheel
#11 369.3   running build
#11 369.3   running build_py
#11 369.3   Traceback (most recent call last):
#11 369.3     File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 37, in silent_call
#11 369.3       subprocess.check_output(cmd, stderr=shut_up)
#11 369.3     File "/root/conda/lib/python3.6/subprocess.py", line 356, in check_output
#11 369.3       **kwargs).stdout
#11 369.3     File "/root/conda/lib/python3.6/subprocess.py", line 438, in run
#11 369.3       output=stdout, stderr=stderr)
#11 369.3   subprocess.CalledProcessError: Command '['cmake', '../compile']' returned non-zero exit status 1.
#11 369.3
#11 369.3   During handling of the above exception, another exception occurred:
#11 369.3
#11 369.3   Traceback (most recent call last):
#11 369.3     File "<string>", line 1, in <module>
#11 369.3     File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 126, in <module>
#11 369.3       url='https://github.com/aksnzhy/xlearn')
#11 369.3     File "/root/conda/lib/python3.6/site-packages/setuptools/__init__.py", line 145, in setup
#11 369.3       return distutils.core.setup(**attrs)
#11 369.3     File "/root/conda/lib/python3.6/distutils/core.py", line 148, in setup
#11 369.3       dist.run_commands()
#11 369.3     File "/root/conda/lib/python3.6/distutils/dist.py", line 955, in run_commands
#11 369.3       self.run_command(cmd)
#11 369.3     File "/root/conda/lib/python3.6/distutils/dist.py", line 974, in run_command
#11 369.3       cmd_obj.run()
#11 369.3     File "/root/conda/lib/python3.6/site-packages/wheel/bdist_wheel.py", line 192, in run
#11 369.3       self.run_command('build')
#11 369.3     File "/root/conda/lib/python3.6/distutils/cmd.py", line 313, in run_command
#11 369.3       self.distribution.run_command(command)
#11 369.3     File "/root/conda/lib/python3.6/distutils/dist.py", line 974, in run_command
#11 369.3       cmd_obj.run()
#11 369.3     File "/root/conda/lib/python3.6/distutils/command/build.py", line 135, in run
#11 369.3       self.run_command(cmd_name)
#11 369.3     File "/root/conda/lib/python3.6/distutils/cmd.py", line 313, in run_command
#11 369.3       self.distribution.run_command(command)
#11 369.3     File "/root/conda/lib/python3.6/distutils/dist.py", line 974, in run_command
#11 369.3       cmd_obj.run()
#11 369.3     File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 104, in run
#11 369.3       compile_cpp()
#11 369.3     File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 75, in compile_cpp
#11 369.3       silent_call(cmake_cmd, raise_error=True, error_msg='Please install CMake first')
#11 369.3     File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 41, in silent_call
#11 369.3       raise Exception(error_msg);
#11 369.3   Exception: Please install CMake first
#11 369.3   ----------------------------------------
#11 369.3   ERROR: Failed building wheel for xlearn
#11 369.3 ERROR: schema 0.7.1 has requirement contextlib2==0.5.5, but you'll have contextlib2 0.6.0.post1 which is incompatible.
#11 369.3     ERROR: Command errored out with exit status 1:
#11 369.3      command: /root/conda/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-_5jp1bcj/xlearn/setup.py'"'"'; __file__='"'"'/tmp/pip-install-_5jp1bcj/xlearn/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-gi2gsk_k/install-record.txt --single-version-externally-managed --compile
#11 369.3          cwd: /tmp/pip-install-_5jp1bcj/xlearn/
#11 369.3     Complete output (51 lines):
#11 369.3     /root/conda/lib/python3.6/site-packages/setuptools/dist.py:476: UserWarning: Normalizing '0.40.a1' to '0.40a1'
#11 369.3       normalized_version,
#11 369.3     running install
#11 369.3     running build
#11 369.3     running build_py
#11 369.3     Traceback (most recent call last):
#11 369.3       File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 37, in silent_call
#11 369.3         subprocess.check_output(cmd, stderr=shut_up)
#11 369.3       File "/root/conda/lib/python3.6/subprocess.py", line 356, in check_output
#11 369.3         **kwargs).stdout
#11 369.3       File "/root/conda/lib/python3.6/subprocess.py", line 438, in run
#11 369.3         output=stdout, stderr=stderr)
#11 369.3     subprocess.CalledProcessError: Command '['cmake', '../compile']' returned non-zero exit status 1.
#11 369.3
#11 369.3     During handling of the above exception, another exception occurred:
#11 369.3
#11 369.3     Traceback (most recent call last):
#11 369.3       File "<string>", line 1, in <module>
#11 369.3       File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 126, in <module>
#11 369.3         url='https://github.com/aksnzhy/xlearn')
#11 369.3       File "/root/conda/lib/python3.6/site-packages/setuptools/__init__.py", line 145, in setup
#11 369.3         return distutils.core.setup(**attrs)
#11 369.3       File "/root/conda/lib/python3.6/distutils/core.py", line 148, in setup
#11 369.3         dist.run_commands()
#11 369.3       File "/root/conda/lib/python3.6/distutils/dist.py", line 955, in run_commands
#11 369.3         self.run_command(cmd)
#11 369.3       File "/root/conda/lib/python3.6/distutils/dist.py", line 974, in run_command
#11 369.3         cmd_obj.run()
#11 369.3       File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 93, in run
#11 369.3         install.run(self)
#11 369.3       File "/root/conda/lib/python3.6/site-packages/setuptools/command/install.py", line 61, in run
#11 369.3         return orig.install.run(self)
#11 369.3       File "/root/conda/lib/python3.6/distutils/command/install.py", line 545, in run
#11 369.3         self.run_command('build')
#11 369.3       File "/root/conda/lib/python3.6/distutils/cmd.py", line 313, in run_command
#11 369.3         self.distribution.run_command(command)
#11 369.3       File "/root/conda/lib/python3.6/distutils/dist.py", line 974, in run_command
#11 369.3         cmd_obj.run()
#11 369.3       File "/root/conda/lib/python3.6/distutils/command/build.py", line 135, in run
#11 369.3         self.run_command(cmd_name)
#11 369.3       File "/root/conda/lib/python3.6/distutils/cmd.py", line 313, in run_command
#11 369.3         self.distribution.run_command(command)
#11 369.3       File "/root/conda/lib/python3.6/distutils/dist.py", line 974, in run_command
#11 369.3         cmd_obj.run()
#11 369.3       File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 104, in run
#11 369.3         compile_cpp()
#11 369.3       File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 75, in compile_cpp
#11 369.3         silent_call(cmake_cmd, raise_error=True, error_msg='Please install CMake first')
#11 369.3       File "/tmp/pip-install-_5jp1bcj/xlearn/setup.py", line 41, in silent_call
#11 369.3         raise Exception(error_msg);
#11 369.3     Exception: Please install CMake first
#11 369.3     ----------------------------------------
#11 369.3 ERROR: Command errored out with exit status 1: /root/conda/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-_5jp1bcj/xlearn/setup.py'"'"'; __file__='"'"'/tmp/pip-install-_5jp1bcj/xlearn/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-gi2gsk_k/install-record.txt --single-version-externally-managed --compile Check the logs for full command output.
#11 369.3
------
failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c conda env update -f base.yaml &&     conda clean -fay &&     python -m ipykernel install --user --name 'python3' --display-name 'python3']: runc did not terminate sucessfully
```

**Comment by gramhagen** [comment]
Seems like conda is not completing channel installation of cmake prior to pip installation of xlearn. 😔 There's still an open issue on building a conda feedstock for xlearn. In the meantime we can revert the removal of explicit cmake installation in the dockerfile.

@yueguoguo should we remove cmake and xlearn from the conda env script and add instructions to build it separately? I'm

**Comment by yueguoguo** [comment]
Maybe we can still keep xlearn in the conda env but install cmake separately.

**Comment by gramhagen** [comment]
sure, let's just drop cmake from conda, add it to docker and then we can add a note in the setup readme indicating xlearn requires cmake.

**Comment by yueguoguo** [comment]
Resolved in #1034
